torch/_dynamo/variables/builder.py

```diff
@@ -2655,6 +2655,8 @@ def update_dim2constraint(dim, constraint_range, name):
             dynamic_size = DimDynamic.DYNAMIC
         elif static_shapes or config.assume_static_by_default or marked_static:
             dynamic_size = DimDynamic.STATIC
+        elif i == 0 and torch._dynamo.config.dynamic_first_dimension:
```

Comment:
You can't do this unconditionally. If I have a transposed tensor like torch.randn(4, 5).T (which has size (5, 4), stride (1, 5), I don't want to mark dim 0 as dynamic because this would result in a size (s0, 4), stride (1, s0) which will destroy perf. A very simple heuristic is to only do this when the input is contiguous, but a better strategy is to mark any dimension dynamic when it is *outermost*, no matter where it is in the tensor.